CoreFoundation/NumberDate.subproj/CFTimeZone.c

```diff
@@ -183,11 +185,133 @@ static void __CFTimeZoneGetOffset(CFStringRef timezone, int32_t *offset) {
 
     RegCloseKey(hKey);
 }
+#elif TARGET_OS_ANDROID
+typedef Boolean (*__CFAndroidTimeZoneListEnumerateCallback)(const char name[40], int32_t start, int32_t length, FILE *fp, void *context);
+
+static void __CFAndroidTimeZoneListEnumerate(__CFAndroidTimeZoneListEnumerateCallback callback, void *context) {
+    // The best reference should be Android Bionic's libc/tzcode/bionic.cpp
+    static const char *tzDataFiles[] = {
+        "/data/misc/zoneinfo/current/tzdata",
+        "/system/usr/share/zoneinfo/tzdata"
+    };
+
+    Boolean done = FALSE;
+    for (int idx = 0; idx < sizeof(tzDataFiles)/sizeof(tzDataFiles[0]) && !done; idx++) {
+        FILE *fp = fopen(tzDataFiles[idx], "rb");
+        if (!fp) {
+            continue;
+        }
+
+        char header[24];
+        if (fread(header, 1, sizeof(header), fp) != sizeof(header)) {
+            fclose(fp);
+            continue;
+        }
+        if (strncmp(header, "tzdata", 6) != 0) {
+            fclose(fp);
+            continue;
+        }
+
+        int32_t indexOffset;
+        memcpy(&indexOffset, &header[12], sizeof(int32_t));
+        indexOffset = betoh32(indexOffset);
+
+        int32_t dataOffset;
+        memcpy(&dataOffset, &header[16], sizeof(int32_t));
+        dataOffset = betoh32(dataOffset);
+
+        if (indexOffset < 0 || dataOffset < indexOffset) {
+            fclose(fp);
+            continue;
+        }
+        if (fseek(fp, indexOffset, SEEK_SET) != 0) {
+            fclose(fp);
+            continue;
+        }
```

Comment:
Can you extract the parsing into a helper function?  That avoids the `fclose(fp); continue;` all over the place.

Reply:
I will see if that's possible, but I’m going to remove a lot of `fclose(fp); continue;` and replace them with a lot of `fclose(fp); return FALSE;`.